### ISSUE TITLE
Elaborate on "deprecating" requestAnimationFrame()

### ISSUE THREAD

**Issue by annevk** [WebKit/explainers#85]
I'd love to understand better as to why extensions to `requestAnimationFrame()` are not the way forward.

Re: https://github.com/WebKit/explainers/tree/main/animation-frame-rate.

**Comment by graouts** [comment]
First, I think it's important to consider that we're not suggesting the deprecation of `requestAnimationFrame()`. That function is part of the Web platform and here to stay.

This proposal is about controlling the frame rate of _animations_. While `requestAnimationFrame()` may be used to yield an animation through repeated calls, it does not in itself provide any animation feature. It has no notion of timing, duration, iteration, easing, progress, etc. This is the province of technologies based on the Web Animations model.

A frame rate makes sense for something that is, by nature, progressing, whereas `requestAnimationFrame()` registers a one-shot callback.

Separately to this explainer, we're advocating the support for [custom effects](https://github.com/w3c/csswg-drafts/issues/6861) to offer support for callback-based animations, which would tie in nicely with this proposal since this would allow authors to write such animations while specifying a frame rate.

**Comment by annevk** [comment]
All true (though I don't personally equate deprecating with removal), but assuming custom effects come to pass, what use cases would remain for `requestAnimationFrame()`?

**Comment by annevk** [comment]
Right, so the idea is that you'd continue to use `requestAnimationFrame()` if you need something to happen at that point in time, but it's not expected you'd use it at 60Hz to move something around on the screen. I guess that's reasonable.

**Comment by mdrejhon** [comment]
TestUFO is one of the world's most popular high-refresh-rate requestAnimationFrame() website.  

www.testufo.com is visited by over 1 million users per month whom run at refresh rates above 60 Hz.  

It is used by hundreds (possibly thousands) of YouTubers, bloggers, review sites (some as prestigious as TomsHardware, CNet, RTINGS, and the scientific researchers listed at www.blurbusters.com/area51 since TestUFO is cited in more than 20 peer reviewed science papers because of its scientifically-accurate refresh rate detection and frame-perfect animations)

Years ago, I had developed heuristics to detect the refresh rate from the tick-tock of requestAnimationFrame() and it successfully works on all major browsers including Chome, FireFox, Opera, Edge, and even Android browsers (of high-Hz phones like Razer)

TestUFO requires:

1. Ability to detect refresh rates.  This is done by a heuristical formula.

2. Ability to detect if refresh rate is accurate or faked.  This is done by a heuristical formula, combined with a simultaneous useragent blacklist & whitelist

3. Ability to self-detect stutters.  This is done by comparing the timestamp of the current frame versus the expected frame.   This is the only way to produce science papers, by guaranteeing that a stutter was detected (and a message displayed).   This is very important for frameskipping tests as well as the "display motion blur" pursuit camera I peer-reviewed paper with NIST.gov, Nokia, and Keltek -- (indie invention that turns a $30,000 science rig into something that can be tested by end users) -- and is used by dozens of YouTubers.  (info: www.blurbusters.com/pursuit-camera 
 .....) and is a major part of RTINGS.com tests as well as other 10-millon-viewer websites/youtubers

5. Combined, all the websites that use Blur Busters inventions (including the TestUFO website that uses requestAnimationFrame() ...), are viewed by over 250 million viewers per month, as Blur Busters is an incubator of indie display-testing inventions for the industry. Even if the users don't view TestUFO directly, the reviewers themselves and the scientists themselves trust & use TestUFO.

I am very concerned that Web Animations API does not satisfy the requirements of the scientists I'm in the peer reviewed papers at www.blurbusters.com/research-papers

**Comment by mdrejhon** [comment]
## We Need An Accurate Frame Presentation Callback Event added to Web Animation API

I have done some research why we are unable to port TestUFO away from requestAnimationFrame(), towards the new Web Animation API (still lower scientific quality for our needs, explained at  https://github.com/WebKit/explainers/issues/89#issuecomment-1192425504

I am pleased to say there's only one change to Web Animation API that we will need -- a callback API for all successfully presented web animation frames.

TestUFO is in several peer-reviewed papers (including one co-authored by NIST.gov) because I've successfully used requestAnimationFrame() as a single-framedrop detector.  

Via analyzing the now() timing of requestAnimationFrame callbacks, I can detect single missed refresh cycles (tiny microstutters, sometimes only visible to a camera).  I know when I'm only running 59fps at 60Hz, or 239fps at 240Hz, and can instantly invalidate a scientific test on time.

To replace requestAnimationFrame() I need a modification to the Web Animation API to add a callback purely for framedrop analysis purpose.  I also will settle for a callback upon missed frames (at a speed of within 1/Hz, so I can render an error message as quickly as possible, for photographic tests of TestUFO animations).

More info why I need a frame presentation time callback:
https://github.com/WebKit/explainers/issues/89#issuecomment-1192425504

**Comment by mdrejhon** [comment]
Also:

Another reason is we automatically use multiple divisors of Hz for many educational tests.  So on a 165 Hz monitor, some animations will play at exactly half 165Hz (87.5fps) or exactly third of 165 Hz (55fps), to compare against full frame rate, e.g.  https://www.testufo.com/blackframes#count=3&bonusufo=1&equalizer=1&background=000000&multistrobe=1&pps=960

Remember, TestUFO has over 30 tests (selectable at upper right corner) with millions of customizable parameters for each (totalled).  

So to cherrypick one example (more education-related rather than researcher-related):

Web Animation API has no easy method to port this specific test -- www.testufo.com/blackframes (or a fancy configured version like testufo.com/blackframes#count=3&bonusufo=1&equalizer=1&background=000000&multistrobe=1&pps=960 -- every combination is fully bookmarkable and shareable, which is a neat feature of TestUFO for education's sake) -- and accurate framedrop-free detection, since the BFI animation erratically flickers if there's even just 1 framedrop, so some motion algorithms need perfect framerate=Hz to be properly educational.   

By utilizing my framedrop detector I use for requestAnimationFrame() I can tell whether the frame rate deviates from perfect framerate=Hz, I can instantly tell the user to close apps/tabs until the animation runs again at full frame rate -- try it and try forcing stutters.  You will see the bottom message change into orange or red, telling users.

Also, we have held off the mobile-compatible version of TestUFO because of Apple's lack of ability to provide an acceptable route for TestUFO.   

The lack of a frame-presented callback event in Web Animation API is highly problematic for a lot of needs of TestUFO, and eventually popup messages at TestUFO will soon display a globally blacklisted-warning for all Safari browsers (never shows a green "READY"), once 120Hz Apple devices are more popular than 60Hz Apple devices (in the next 5 years or so).  

## We urgently need to fix the Web Animation API limitation, since it is not possible to deprecate requestAnimationFrame() otherwise
